File: Android/app/src/main/java/com/hhp227/paging3/paging/LoadStates.kt
package com.hhp227.paging3.paging

// 원본과 동일
data class LoadStates(
    val refresh: LoadState,
    val prepend: LoadState,
    val append: LoadState
) {
    inline fun forEach(op: (LoadType, LoadState) -> Unit) {
        op(LoadType.REFRESH, refresh)
        op(LoadType.PREPEND, prepend)
        op(LoadType.APPEND, append)
    }

    internal fun modifyState(loadType: LoadType, newState: LoadState): LoadStates {
        return when (loadType) {
            LoadType.APPEND -> copy(
                append = newState
            )
            LoadType.PREPEND -> copy(
                prepend = newState
            )
            LoadType.REFRESH -> copy(
                refresh = newState
            )
        }
    }

    internal fun get(loadType: LoadType) = when (loadType) {
        LoadType.REFRESH -> refresh
        LoadType.APPEND -> append
        LoadType.PREPEND -> prepend
    }

    internal companion object {
        val IDLE = LoadStates(
            refresh = LoadState.NotLoading(false),
            prepend = LoadState.NotLoading(false),
            append = LoadState.NotLoading(false)
        )
    }
}

File: iOS/Paging/Paging/LoadStates.swift
//
//  LoadStates.swift
//  Paging
//
//  Created by 홍희표 on 2022/06/19.
//

struct LoadStates: Equatable {
    let refresh: LoadState
    
    let prepend: LoadState
    
    let append: LoadState

    func forEach(_ op: (LoadType, LoadState) -> Void) {
        op(.REFRESH, refresh)
        op(.PREPEND, prepend)
        op(.APPEND, append)
    }

    func modifyState(_ loadType: LoadType, _ newState: LoadState) -> LoadStates {
        switch loadType {
        case .APPEND:
            return LoadStates(
                refresh: self.refresh,
                prepend: self.prepend,
                append: newState
            )
        case .PREPEND:
            return LoadStates(
                refresh: self.refresh,
                prepend: newState,
                append: self.append
            )
        case .REFRESH:
            return LoadStates(
                refresh: newState,
                prepend: self.prepend,
                append: self.append
            )
        }
    }
    
    func get(_ loadType: LoadType) -> LoadState {
        switch loadType {
        case .REFRESH:
            return self.refresh
        case .APPEND:
            return self.append
        case .PREPEND:
            return self.prepend
        }
    }
    
    func toString() -> String {
        return "LoadStates(refresh=\(refresh), prepend=\(prepend), append=\(append)"
    }
    
    static let IDLE = LoadStates(
        refresh: LoadState.NotLoading(false),
        prepend: LoadState.NotLoading(false),
        append: LoadState.NotLoading(false)
    )
}
